File: app/src/main/java/com/example/techfix/AppointmentDetailActivity.java
package com.example.techfix;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;

import com.example.techfix.data.AppointmentRepository;
import com.example.techfix.data.BranchRepository;
import com.example.techfix.data.FirestoreSingleCallback;
import com.example.techfix.data.PaymentRepository;
import com.example.techfix.data.RepairServiceRepository;
import com.example.techfix.model.Appointment;
import com.example.techfix.model.Branch;
import com.example.techfix.model.Payment;
import com.example.techfix.model.RepairService;

import java.io.File;
import java.io.IOException;
import java.util.Locale;

public class AppointmentDetailActivity extends AppCompatActivity {

    private AppointmentRepository appointmentRepository;
    private PaymentRepository paymentRepository;
    private String appointmentId;
    private double servicePrice;
    private Uri currentPhotoUri;

    private final androidx.activity.result.ActivityResultLauncher<String> cameraPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), granted -> {
                if (granted) launchCamera();
                else Toast.makeText(this, "Camera permission is required", Toast.LENGTH_LONG).show();
            });

    private final androidx.activity.result.ActivityResultLauncher<Uri> takePictureLauncher =
            registerForActivityResult(new ActivityResultContracts.TakePicture(), success -> {
                if (success && currentPhotoUri != null) {
                    ((ImageView) findViewById(R.id.ivRepairPhoto)).setImageURI(currentPhotoUri);
                    appointmentRepository.updateImage(appointmentId, currentPhotoUri.toString());
                    Toast.makeText(this, "Photo attached", Toast.LENGTH_SHORT).show();
                }
            });

    private final androidx.activity.result.ActivityResultLauncher<Intent> paymentGatewayLauncher =
            registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
                if (result.getResultCode() == RESULT_OK) {
                    updatePaymentSection();
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_appointment_detail);

        appointmentRepository = new AppointmentRepository();
        paymentRepository = new PaymentRepository();
        appointmentId = getIntent().getStringExtra("APPOINTMENT_ID");

        loadAppointmentDetails();

        findViewById(R.id.btnTakePhoto).setOnClickListener(v -> checkCameraPermissionAndLaunch());
        findViewById(R.id.btnPayNow).setOnClickListener(v -> {
            Intent intent = new Intent(AppointmentDetailActivity.this, PaymentGatewayActivity.class);
            intent.putExtra("APPOINTMENT_ID", appointmentId);
            intent.putExtra("AMOUNT", servicePrice);
            paymentGatewayLauncher.launch(intent);
        });
    }

    private void loadAppointmentDetails() {
        appointmentRepository.getAppointmentById(appointmentId, new FirestoreSingleCallback<Appointment>() {
            @Override
            public void onSuccess(Appointment appointment) {
                BranchRepository branchRepository = new BranchRepository();
                RepairServiceRepository serviceRepository = new RepairServiceRepository();

                branchRepository.getBranchById(appointment.branchId, new FirestoreSingleCallback<Branch>() {
                    @Override
                    public void onSuccess(Branch branch) {
                        ((TextView) findViewById(R.id.tvDetailBranch)).setText(branch.name);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        ((TextView) findViewById(R.id.tvDetailBranch)).setText("Branch");
                    }
                });

                serviceRepository.getServiceById(appointment.serviceId, new FirestoreSingleCallback<RepairService>() {
                    @Override
                    public void onSuccess(RepairService service) {
                        servicePrice = service.price;
                        ((TextView) findViewById(R.id.tvDetailService)).setText(service.name);
                        updatePaymentSection();
                    }

                    @Override
                    public void onFailure(Exception e) {
                        ((TextView) findViewById(R.id.tvDetailService)).setText("Service");
                    }
                });

                ((TextView) findViewById(R.id.tvDetailDate)).setText(appointment.requestedDate);
                ((TextView) findViewById(R.id.tvDetailStatus)).setText(appointment.status);

                if (appointment.imageUri != null && !appointment.imageUri.isEmpty()) {
                    ((ImageView) findViewById(R.id.ivRepairPhoto)).setImageURI(Uri.parse(appointment.imageUri));
                }
            }

            @Override
            public void onFailure(Exception e) {
                Toast.makeText(AppointmentDetailActivity.this, "Failed to load appointment", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void updatePaymentSection() {
        TextView paymentStatusView = findViewById(R.id.tvPaymentStatus);
        Button payButton = findViewById(R.id.btnPayNow);

        paymentRepository.getPaymentByAppointment(appointmentId, new FirestoreSingleCallback<Payment>() {
            @Override
            public void onSuccess(Payment payment) {
                if ("Paid".equals(payment.status)) {
                    paymentStatusView.setText("Paid on " + payment.paymentDate + " via " + payment.method);
                    payButton.setEnabled(false);
                    payButton.setText("Already Paid");
                } else {
                    paymentStatusView.setText(String.format(Locale.getDefault(), "Rs. %.2f — Not Paid", servicePrice));
                }
            }

            @Override
            public void onFailure(Exception e) {
                // No payment record exists yet — normal for a fresh appointment.
                paymentStatusView.setText(String.format(Locale.getDefault(), "Rs. %.2f — Not Paid", servicePrice));
            }
        });
    }

    private void checkCameraPermissionAndLaunch() {
        boolean hasPermission = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED;
        if (hasPermission) {
            launchCamera();
        } else {
            cameraPermissionLauncher.launch(Manifest.permission.CAMERA);
        }
    }

    private void launchCamera() {
        try {
            File photoFile = File.createTempFile(
                    "repair_" + appointmentId + "_",
                    ".jpg",
                    getExternalFilesDir("Pictures")
            );
            currentPhotoUri = FileProvider.getUriForFile(
                    this,
                    getPackageName() + ".fileprovider",
                    photoFile
            );
            takePictureLauncher.launch(currentPhotoUri);
        } catch (IOException e) {
            Toast.makeText(this, "Couldn't create photo file: " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/techfix/AppointmentDetailActivity.kt
package com.example.techfix

import android.Manifest
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.lifecycle.lifecycleScope
import com.example.techfix.data.AppDatabase
import com.example.techfix.data.entity.Payment
import kotlinx.coroutines.launch
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class AppointmentDetailActivity : AppCompatActivity() {

    private lateinit var db: AppDatabase
    private var appointmentId: Int = -1
    private var servicePrice: Double = 0.0
    private var currentPhotoUri: Uri? = null

    private val cameraPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted) launchCamera() else Toast.makeText(this, "Camera permission is required", Toast.LENGTH_LONG).show()
    }

    private val takePictureLauncher = registerForActivityResult(
        ActivityResultContracts.TakePicture()
    ) { success ->
        if (success && currentPhotoUri != null) {
            findViewById<ImageView>(R.id.ivRepairPhoto).setImageURI(currentPhotoUri)
            lifecycleScope.launch {
                db.appointmentDao().updateImage(appointmentId, currentPhotoUri.toString())
            }
            Toast.makeText(this, "Photo attached", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_appointment_detail)

        db = AppDatabase.getDatabase(applicationContext)
        appointmentId = intent.getIntExtra("APPOINTMENT_ID", -1)

        loadAppointmentDetails()

        findViewById<Button>(R.id.btnTakePhoto).setOnClickListener {
            checkCameraPermissionAndLaunch()
        }

        findViewById<Button>(R.id.btnPayNow).setOnClickListener {
            processPayment()
        }
    }

    private fun loadAppointmentDetails() {
        lifecycleScope.launch {
            val appointment = db.appointmentDao().getAppointmentById(appointmentId) ?: return@launch
            val branch = db.branchDao().getBranchById(appointment.branchId)
            val service = db.repairServiceDao().getServiceById(appointment.serviceId)
            servicePrice = service?.price ?: 0.0

            findViewById<TextView>(R.id.tvDetailService).text = service?.name ?: "Service"
            findViewById<TextView>(R.id.tvDetailBranch).text = branch?.name ?: "Branch"
            findViewById<TextView>(R.id.tvDetailDate).text = appointment.requestedDate
            findViewById<TextView>(R.id.tvDetailStatus).text = appointment.status

            appointment.imageUri?.let { uriString ->
                findViewById<ImageView>(R.id.ivRepairPhoto).setImageURI(Uri.parse(uriString))
            }

            val payment = db.paymentDao().getPaymentByAppointment(appointmentId)
            val paymentStatusView = findViewById<TextView>(R.id.tvPaymentStatus)
            val payButton = findViewById<Button>(R.id.btnPayNow)
            if (payment != null && payment.status == "Paid") {
                paymentStatusView.text = "Paid on ${payment.paymentDate} via ${payment.method}"
                payButton.isEnabled = false
                payButton.text = "Already Paid"
            } else {
                paymentStatusView.text = "Rs. %.2f — Not Paid".format(servicePrice)
            }
        }
    }

    private fun checkCameraPermissionAndLaunch() {
        val hasPermission = ContextCompat.checkSelfPermission(
            this, Manifest.permission.CAMERA
        ) == PackageManager.PERMISSION_GRANTED

        if (hasPermission) {
            launchCamera()
        } else {
            cameraPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    private fun launchCamera() {
        val photoFile = File.createTempFile(
            "repair_${appointmentId}_",
            ".jpg",
            getExternalFilesDir("Pictures")
        )
        currentPhotoUri = FileProvider.getUriForFile(
            this,
            "${packageName}.fileprovider",
            photoFile
        )
        takePictureLauncher.launch(currentPhotoUri)
    }

    private fun processPayment() {
        lifecycleScope.launch {
            val existing = db.paymentDao().getPaymentByAppointment(appointmentId)
            val today = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())

            if (existing == null) {
                db.paymentDao().insertPayment(
                    Payment(
                        appointmentId = appointmentId,
                        amount = servicePrice,
                        status = "Paid",
                        paymentDate = today,
                        method = "Cash"
                    )
                )
            } else {
                db.paymentDao().markAsPaid(existing.paymentId, "Paid", today)
            }

            Toast.makeText(this@AppointmentDetailActivity, "Payment successful!", Toast.LENGTH_SHORT).show()
            loadAppointmentDetails()
        }
    }
}